Replace RSA references with more generic public key API

alexeblee opened the pull request:
Related to #537, so it may be helpful to review that change first. 

I've removed most hardcoded references to RSA and replaced them with a generic s2n_pkey API that handles public key crypto operations without referencing a specific algorithm. Currently, s2n_pkey has wrapper functions that call algorithm-specific functions defined in crypto/s2n_rsa and crypto/s2n_ecdsa. There were a few different representations of "keys" sprinkled throughout (in s2n_ecc.h and s2n_certificate.h, for example), and I've tried to consolidate them where I thought it made sense.

raycoll commented on crypto/s2n_pkey.h:
```diff
@@ -0,0 +1,43 @@
+    const struct s2n_pkey_ctx *ctx;
+};
+
+struct s2n_pkey_ctx {
```
nit: usually `*_ctx` structs are containers of data. Any reason why these function handlers need to be in a separate struct? We could just add them as fields in the `s2n_pkey` struct.

alexeblee replied:
I've collapsed them into the single s2n_pkey struct and tweaked how these function pointers are set. Let me know if this looks better to you.

raycoll commented on crypto/s2n_pkey.c:
```diff
@@ -0,0 +1,136 @@
+
+int s2n_asn1der_to_private_key(struct s2n_pkey *priv_key, struct s2n_blob *asn1der)
+{
+    int ret;
```
nit: This is a bit more readable to me if we move this declaration down to where it's first used.

raycoll commented on crypto/s2n_pkey.c:
```diff
@@ -0,0 +1,136 @@
+
+int s2n_asn1der_to_public_key(struct s2n_pkey *pub_key, struct s2n_blob *asn1der)
+{
+    int ret;
```
Same here

raycoll commented on crypto/s2n_pkey.c:
```diff
@@ -0,0 +1,136 @@
+#include <openssl/evp.h>
```
copyright

raycoll commented on crypto/s2n_pkey.h:
```diff
@@ -0,0 +1,43 @@
+#pragma once
```
copyright

bpdavidson commented on tls/s2n_config.c:
```diff
@@ -229,7 +229,7 @@ int s2n_config_add_cert_chain_and_key(struct s2n_config *config, const char *cer
     config->cert_and_key_pairs = (struct s2n_cert_chain_and_key *)(void *)mem.data;
     config->cert_and_key_pairs->head = NULL;
-    config->cert_and_key_pairs->private_key.rsa = NULL;
+    config->cert_and_key_pairs->private_key.ctx = NULL;
```
Why not also preserve the removed line with:
`config->cert_and_key_pairs->private_key.rsa_key.rsa = NULL;`

alexeblee commented on crypto/s2n_pkey.c:
```diff
@@ -0,0 +1,136 @@
+    /* Initialize s2n_pkey according to key type */
+    switch (type) {
+    case EVP_PKEY_RSA:
+        ret = s2n_pkey_to_rsa_private_key(&priv_key->key.rsa_key, pkey);
```
It's not ideal, but I wasn't using GUARD here because I still needed to call EVP_PKEY_free at the end outside of the switch statement.

alexeblee commented on crypto/s2n_rsa.c:
```diff
@@ -24,89 +23,56 @@
-        EVP_PKEY_free(public_key);
+int s2n_pkey_to_rsa_public_key(s2n_rsa_public_key *rsa_key, EVP_PKEY *pkey)
+{
+    RSA *rsa = EVP_PKEY_get1_RSA(pkey);
```
The variable naming for rsa and ec_key were based off of the OpenSSL type. I've updated the pkey to say evp_public/private_key though.

alexeblee commented on crypto/s2n_pkey.c:
```diff
@@ -0,0 +1,136 @@
+        priv_key->ctx = &rsa_key_ctx;
+        break;
+    case EVP_PKEY_EC:
+        ret = s2n_pkey_to_ecdsa_private_key(&priv_key->key.ecdsa_key, pkey);
```
The functions were used as helper-ish functions to convert between specific types (s2n_rsa_public_key, s2n_ecdsa_public_key) and don't take the same parameters. I tried to keep the s2n_pkey functions more scoped to what you can conceptually do with a public/private key.

alexeblee commented on crypto/s2n_rsa.c:
```diff
@@ -24,89 +23,56 @@
-
-    if (!RSA_check_key(rsa_key)) {
+    
+    if (!RSA_check_key(rsa)) {
```
It would take a little more finessing to get the correct type for the function argument and to create a generic function to satisfy both use cases. `s2n_rsa_check_key` is called to verify that the certificate parsing/validation function correctly set up the `s2n_pkey` object (which contains an `RSA` object). `RSA_check_key` is called on an `RSA` object inside of `s2n_pkey_to_rsa_private_key` to determine if the `RSA` object is valid before assigning it to an `s2n_rsa_private_key` object.

alexeblee commented on crypto/s2n_rsa.c:
```diff
@@ -124,15 +90,15 @@ static int s2n_rsa_modulus_check(RSA *rsa)
 }
 
-int s2n_rsa_private_encrypted_size(struct s2n_rsa_private_key *key)
+int s2n_rsa_private_encrypted_size(const s2n_rsa_private_key *key)
```
These were originally written as distinct functions probably to make it more readable for the callers. I think they also reflect conceptually what is happening (encryption with the public key vs encryption with the private key). I've made a common static function that both use to mitigate your concern about code duplication.

alexeblee commented on crypto/s2n_certificate.h:
```diff
@@ -53,30 +53,26 @@ typedef enum {
 } s2n_cert_type;
 
-struct s2n_cert_public_key {
+struct s2n_cert {
```
I've reverted the renaming. It still kind of nags me that the cert type is bundled up in a "key type", but I plan on addressing that plus some other refactoring in a separate PR.

alexeblee commented on crypto/s2n_ecdsa.h:
```diff
@@ -24,22 +24,23 @@
 
-extern int s2n_ecdsa_keys_match(const struct s2n_ecdsa_public_key *pub_key, const struct s2n_ecdsa_private_key *priv_key); 
+extern int s2n_pkey_to_ecdsa_public_key(s2n_ecdsa_public_key *ecdsa_key, EVP_PKEY *pkey);
+extern int s2n_pkey_to_ecdsa_private_key(s2n_ecdsa_private_key *ecdsa_key, EVP_PKEY *pkey);
```
Agreed. I've changed these to s2n_evp_pkey_to*

alexeblee commented on crypto/s2n_ecdsa.c:
```diff
@@ -28,9 +28,22 @@
+
+int s2n_ecdsa_sign(const struct s2n_pkey *priv, struct s2n_hash_state *digest, struct s2n_blob *signature)
 {
+    s2n_ecdsa_private_key key = priv->key.ecdsa_key;
```
Fixed in all places.

alexeblee commented:
I think I've addressed all of the feedback from @bpdavidson and @raycoll. I didn't comment on every single comment, but tried to respond to those that required more than just a "Fixed."

alexeblee commented on crypto/s2n_rsa.h:
```diff
@@ -15,31 +15,35 @@
 
-struct s2n_rsa_private_key {
+struct s2n_rsa_key {
     RSA *rsa;
```
I think the original intent was to match the openssl naming scheme. I've found a lot of code that looks something like this:
```
struct s2n_rsa_key rsa_key;
...
rsa_key.rsa
```
For these instances, it looks cleaner to me to keep it as `rsa`.

bpdavidson replied:
Ahh, sounds good.

bpdavidson commented on crypto/s2n_rsa.c:
```diff
@@ -273,3 +208,60 @@ int s2n_rsa_keys_match(struct s2n_rsa_public_key *pub, struct s2n_rsa_private_ke
+    return 0;
+}
+
+int s2n_rsa_check_key(const struct s2n_pkey *pkey)
```
From the `RSA_check_key()` doc, this can only be used on `RSA` structs with private key data. As such, I think this should be modified to:
```
int s2n_rsa_check_key(const struct s2n_rsa_private_key *rsa_key)
{
    notnull_check(rsa_key->rsa);
    GUARD(RSA_check_key(rsa_key->rsa));
    return 0;
}
```
Thoughts?

alexeblee replied:
It's been so long since I've written this code, that I don't remember my rationale for writing it this way. It seems like the way it is written wouldn't work at all because, as you said, RSA_check_key should only work on public keys. I've renamed this function to imply that it is simply doing a check that the public key pkey was initialized appropriately. I *think* that is what I originally intended. It is only called in the client auth validation callbacks which requires that the public key is extracted from the cert. Other calls to RSA_check_key (that are called on private keys) remain as they were.

bpdavidson commented on crypto/s2n_rsa.c:
```diff
@@ -273,3 +208,60 @@ int s2n_rsa_keys_match(struct s2n_rsa_public_key *pub, struct s2n_rsa_private_ke
+        S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
+    }
+    
+    if (!RSA_check_key(rsa)) {
```
`s2n_rsa_check_key()` should be used instead if you make the suggested change to it.

alexeblee commented on crypto/s2n_ecdsa.c:
```diff
@@ -71,123 +78,91 @@ int s2n_ecdsa_verify(const struct s2n_ecdsa_public_key *key, struct s2n_hash_sta
-    /* If key parsing is successful, d2i_PrivateKey increments *key_to_parse to the byte following the parsed data */
-    uint32_t parsed_len = key_to_parse - asn1der->data;
-    if (parsed_len != asn1der->size) {
+int s2n_evp_pkey_to_ecdsa_private_key(s2n_ecdsa_private_key *ecdsa_key, EVP_PKEY *evp_private_key)
```
There is a subtle difference in the error messages used in each function. I could converge them and use a more generic "Error decoding key" message fairly easily. I made that change, and went to go fix the equivalent RSA functions, but realized that the RSA functions can't converge because there are also minor functional differences for public and private keys. (RSA_check_key can't be called on public keys). My personal preference would be to leave them as separate functions for now to keep the RSA and ECDSA cases "symmetrical". If you have a strong opinion to the contrary, let me know.

alexeblee commented on crypto/s2n_ecdsa.c:
```diff
@@ -71,123 +78,91 @@ int s2n_ecdsa_verify(const struct s2n_ecdsa_public_key *key, struct s2n_hash_sta
+{
+    struct s2n_ecdsa_key *ecdsa_key = &pkey->key.ecdsa_key;
+    notnull_check(ecdsa_key->ec_key);
+    EC_KEY_free(ecdsa_key->ec_key);
```
Good catch. That was not intentional.

alexeblee commented on crypto/s2n_pkey.c:
```diff
@@ -0,0 +1,187 @@
+int s2n_pkey_free(struct s2n_pkey *pkey)
+{
+    if (pkey == NULL || pkey->free == NULL) {
+        return 0;
```
Hmm.. I see your point. `s2n_pkey_zero` is actually called just during initialization of the key. For example, in `s2n_config_add_cert_chain_and_key`, we initialize several pointer members to NULL, and `s2n_pkey_zero` was added to achieve that for the key. It's also called when parsing the client certificate. I don't think I can move pkey_zero into pkey_free, but I think I could change the name to more strongly imply it is used to initialize rather than destroy.

alexeblee commented on crypto/s2n_pkey.c:
```diff
@@ -0,0 +1,187 @@
+    if (pkey == NULL || pkey->free == NULL) {
+        return 0;
+    }
+    return pkey->free(pkey);
```
The issue I encountered when I originally did something similar, was that `s2n_connection_new` calls `s2n_connection_wipe` and and tries to wipe `secure.server_public_key` and `secure.client_public_key`. Without digging in to the s2n_pkey struct members, there isn't a great way for that function to determine that the keys have not been initialized. The null check was instead encapsulated inside the s2n_pkey_free function.

I understand the concern of having a state where the key is only half-initialized or half-freed, but I don't know how to guard against a developer making that mistake. Right now, both code paths where a key is initialized or freed correctly handle this.

alexeblee commented on crypto/s2n_pkey.c:
```diff
@@ -0,0 +1,187 @@
+        if ((ret = s2n_rsa_pkey_init(priv_key)) != 0) {
+            break;
+        }
+        ret = s2n_evp_pkey_to_rsa_private_key(&priv_key->key.rsa_key, evp_private_key);
```
I agree that it's really ugly, but I wanted to check the return of s2n_rsa_pkey_init and not call s2n_evp_pkey_to_rsa_private_key if that failed.

alexeblee commented on crypto/s2n_rsa.c:
```diff
@@ -133,20 +53,22 @@ static int s2n_rsa_modulus_check(RSA *rsa)
 }
 
-int s2n_rsa_private_encrypted_size(struct s2n_rsa_private_key *key)
+int s2n_rsa_public_encrypted_size(const s2n_rsa_public_key *key)
```
No functional reason. As per my previous comment, this was just to keep in line with the readability of the code that was originally there. There is overlap between some public and private key operations, but I find it easier to read when it's explicit which key you are operating on.

alexeblee commented on crypto/s2n_rsa.c:
```diff
@@ -273,3 +208,60 @@ int s2n_rsa_keys_match(struct s2n_rsa_public_key *pub, struct s2n_rsa_private_ke
+{
+    struct s2n_rsa_key *rsa_key = &pkey->key.rsa_key;
+
+    RSA_free(rsa_key->rsa);
```
RSA_free has a null check on the first part of the function. I agree that we shouldn't rely on that check though. I've added a null check, but return a "success" even if it's null to keep in line with the idea that you can call free on a null

raycoll commented on stuffer/s2n_stuffer_pem.c:
```diff
@@ -14,7 +14,7 @@
 
 #include <string.h>
-
+#include <stdio.h>
```
Don't think we need this

raycoll commented on crypto/s2n_pkey.c:
```diff
@@ -0,0 +1,187 @@
+    int ret;
+    switch (type) {
+    case EVP_PKEY_RSA:
+        if ((ret = s2n_rsa_pkey_init(pub_key)) != 0) {
```
Let's not do assignment in the conditionals. In my opinion, adding an extra statement to set `ret` before the conditional will help readibility and reduce chance of someone introducing an error here later.

raycoll commented on crypto/s2n_pkey.c:
```diff
@@ -0,0 +1,187 @@
+        ret = s2n_evp_pkey_to_rsa_public_key(&pub_key->key.rsa_key, evp_public_key);
+        break;
+    case EVP_PKEY_EC:
+        if ((ret = s2n_ecdsa_pkey_init(pub_key)) != 0) {
```
Here too
